### PR TITLE
630/add efficientnet lite

### LINKED ISSUE #630: Add EfficientNet Lite model variants.
**Short Description**
EfficientNet Lite models are lighter variants of EfficientNets more suitable for mobile deployment. 

**Existing Implementations**
Original implementation - [link](https://github.com/tensorflow/tpu/tree/master/models/official/efficientnet/lite)
My reimplementation - [link](https://github.com/sebastian-sz/efficientnet-lite-keras)

**Other Information**
EfficientNet Lite's have been requested multiple times in the Tensorflow/Keras ecosystem:

https://github.com/keras-team/keras/issues/16835
https://github.com/tensorflow/tensorflow/issues/45091
https://github.com/tensorflow/tensorflow/pull/48400

Let me know what you think!

### PR DESCRIPTION
**Pull request by sebastian-sz**
# What does this PR do?
Adds EfficientNet Lite variants to `keras_cv` models.

Fixes #630 

This is a port of PR from the [Keras repository](https://github.com/keras-team/keras/pull/16905), as per [this comment].(https://github.com/keras-team/keras/pull/16905#issuecomment-1262811641)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/keras-team/keras-cv/blob/master/.github/CONTRIBUTING.md),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case.
- [x] Did you write any new necessary tests?
- [ ] If this adds a new model, can you run a few training steps on TPU in Colab to ensure that no XLA incompatible OP are used?

## Who can review?
@LukeWood 

Let me know if I should tag anyone else :)

### CODE REVIEW

**Review by LukeWood — CHANGES_REQUESTED** [review]
Great job on the PR @sebastian-sz 

Quick question: do you have any weights for any of these variants?

**Comment by sebastian-sz** [comment]
@LukeWood I do have [weights](https://github.com/sebastian-sz/efficientnet-lite-keras/releases/tag/v1.0) converted from the original tpu repository, but they expect different preprocessing:  
`Normalization(mean=127.0, variance=128.0**2)` instead of Keras-CV `Rescaling(1.0 / 255.0)`.

Using `Rescaling` layer with these weights will probably result in lower Imagenet accuracy.

File: keras_cv/models/efficientnet_lite.py [hunk 1]
```diff
@@ -0,0 +1,649 @@
...
+    """Create EfficientNetLite B4 variant."""
+    return EfficientNetLite(
+        include_rescaling,
+        include_top,
+        width_coefficient=1.4,
+        depth_coefficient=1.8,
+        default_size=300,
+        dropout_rate=0.3,
+        model_name="efficientnetlite4",
+        weights=parse_weights(weights, include_top, "efficientnetlite4"),
```
Comment (ianstenbit) [inline keras_cv/models/efficientnet_lite.py:635]: For all of these (and maybe also model_names) -- let's include the 'b' for consistency with EfficientNetV2 (e.g. `efficientnetliteb4`)

File: keras_cv/models/efficientnet_lite.py [hunk 2]
```diff
@@ -0,0 +1,649 @@
...
+}
+
+BASE_DOCSTRING = """Instantiates the {name} architecture.
+
+    Reference:
+    - [EfficientNet: Rethinking Model Scaling for Convolutional Neural Networks](
+        https://arxiv.org/abs/1905.11946) (ICML 2019)
+
+    This function returns a Keras image classification model,
+    optionally loaded with weights pre-trained on ImageNet.
```
Comment (ianstenbit) [inline keras_cv/models/efficientnet_lite.py:123]: We will eventually pre-train on other datasets, so let's leave out the reference to ImageNet in particular.

(It looks like we include this phrasing in EfficientNet and EfficientNetV2 but we should remove them there too -- feel free to do that in this PR if you'd like to)

File: keras_cv/models/efficientnet_lite.py [hunk 3]
```diff
@@ -0,0 +1,649 @@
...
+    include_top,
+    classes=None,
+    weights=None,
+    input_shape=(None, None, 3),
+    input_tensor=None,
+    pooling=None,
+    classifier_activation="softmax",
+    **kwargs,
+):
+    """Create EfficientNetLite B4 variant."""
```
Comment (ianstenbit) [inline keras_cv/models/efficientnet_lite.py:626]: Since we set these docstrings at the end of the file, we can remove them from inside this method

**Review by ianstenbit — CHANGES_REQUESTED** [review]
Thanks for the PR! We're glad to have EfficientNetLite in KerasCV.

Are you able to run our [training script](https://github.com/keras-team/keras-cv/blob/master/examples/training/classification/imagenet/basic_training.py) to verify that these models (just one is fine for now) converge on ImageNet / potentially provide weights alongside this PR?

If not, I will run the script, but I currently have a backlog of training experiments, so this would be at least a week from now.

**Comment by bhack** [comment]
> If not, I will run the script, but I currently have a backlog of training experiments, so this would be at least a week from now.

I don't know if you could discuss this with the Team but it would be very nice to launch a training job from a PR with a Github Action (after a manual trigger).

It will help us to follow the training on the same page and we could automate the upload of the log to `https://tensorboard.dev/`.

If not I think it will be a little bit hard to scale at some point.

**Comment by ianstenbit** [comment]
> I don't know if you could discuss this with the Team but it would be very nice to launch a training job from a PR with a Github Action (after a manual trigger).
> 
> It will help us to follow the training on the same page and we could automate the upload of the log to https://tensorboard.dev/.
> 
> If not I think it will be a little bit hard to scale at some point.

This is on our radar. It's something we'd like to include eventually, but haven't yet prioritized.

**Comment by bhack** [comment]
> This is on our radar. It's something we'd like to include eventually, but haven't yet prioritized.

Good, when you will be ready I hope that we find a space to discuss this with the community before you finalize some details.

File: keras_cv/models/efficientnet_lite.py (hunk 1 above)
Comment (sebastian-sz) [reply keras_cv/models/efficientnet_lite.py]: Done. Added `b`'s.

File: keras_cv/models/efficientnet_lite.py (hunk 2 above)
Comment (sebastian-sz) [reply keras_cv/models/efficientnet_lite.py]: Done removed the line mentioning weights. Also removed from EfficientNetV1 & V2.

File: keras_cv/models/efficientnet_lite.py (hunk 3 above)
Comment (sebastian-sz) [reply keras_cv/models/efficientnet_lite.py]: Done.

**Comment by sebastian-sz** [comment]
@ianstenbit 
>Are you able to run our [training script](https://github.com/keras-team/keras-cv/blob/master/examples/training/classification/imagenet/basic_training.py) to verify that these models (just one is fine for now) converge on ImageNet / potentially provide weights alongside this PR?

Sorry, I do not have the compute resources to run Imagenet train job :(

**Comment by ianstenbit** [comment]
> @ianstenbit
> 
> > Are you able to run our [training script](https://github.com/keras-team/keras-cv/blob/master/examples/training/classification/imagenet/basic_training.py) to verify that these models (just one is fine for now) converge on ImageNet / potentially provide weights alongside this PR?
> 
> Sorry, I do not have the compute resources to run Imagenet train job :(

No worries at all -- we need a solution to provide GCP resources for training in situations like these. For now, I will try to train one of these in the next week.

**Review by ianstenbit — APPROVED** [review]

**Comment by ianstenbit** [comment]
Thank you! Once I can produce some benchmark training results / weights for one of these I will merge.

**Comment by sebastian-sz** [comment]
@ianstenbit Thanks! Looking forward to the results, as I am curious as well.

**Comment by sebastian-sz** [comment]
@ianstenbit Thank you for the update. 

> I'm using [this](https://paperswithcode.com/model/tf-efficientnet-lite?variant=tf-efficientnet-lite0) as benchmark scores

It seems metrics from timm + papers with code are slightly higher than the ones from [tensorflow/tpu repository](https://github.com/tensorflow/tpu/blob/master/models/official/efficientnet/lite/README.md). 

Looking forward to the results!

**Comment by ianstenbit** [comment]
First training run had ~70.1% top-1 accuracy for EfficientNetLiteB0. I am re-starting a new run with a higher batch size and a warmped-up cosine decay LR schedule and I'll see how that goes. May need to add weight decay as well -- we shall see.

**Review by LukeWood — APPROVED** [review]

**Comment by sebastian-sz** [comment]
@ianstenbit Sorry to hear about the results. If I have more time I will probably look into it too. 

Thanks for the merge!
